akka-distributed-data/src/main/scala/akka/cluster/ddata/VersionVector.scala

```diff
@@ -3,21 +3,37 @@
  */
 package akka.cluster.ddata
 
+import scala.collection.immutable
 import java.util.concurrent.atomic.AtomicLong
-
 import scala.annotation.tailrec
 import scala.collection.immutable.TreeMap
-
 import akka.cluster.Cluster
 import akka.cluster.UniqueAddress
+import akka.cluster.UniqueAddress
 
 /**
  * VersionVector module with helper classes and methods.
  */
 object VersionVector {
 
-  val empty: VersionVector = new VersionVector(TreeMap.empty[UniqueAddress, Long])
+  private val emptyVersions: TreeMap[UniqueAddress, Long] = TreeMap.empty
+  val empty: VersionVector = new VersionVector(Left(emptyVersions))
+
   def apply(): VersionVector = empty
+
+  def apply(versions: TreeMap[UniqueAddress, Long]): VersionVector =
+    if (versions.isEmpty) empty
+    else if (versions.size == 1) apply(versions.head._1, versions.head._2)
+    else new VersionVector(Left(versions))
```

Comment:
Yes, I also think it allocates a tuple. This is the reason I also avoid the unapply style variable assignment in performance sensitive code.